### PR TITLE
Scala math

### PR DESCRIPTION
**Pull request by bob-bins**
Thought I did this last night but I can't find any history of it, so I'm trying again.

Tests at the bottom are a superset of our sinon, chai, and sunday mocha tests for complex numbers. It's quick and dirty since I am assuming setting up the tests properly would be done with you.

### CODE REVIEW

File: lib/public/scala/complex-number.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
+//import Numeric._
+
+class ComplexNumber(var real:Double, var imaginary:Double) {
```
Comment (tommysullivan) [inline lib/public/scala/complex-number.scala]: var == yucky mutability! we can use val

File: lib/public/scala/complex-number.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+  def +(that:Double):ComplexNumber = new ComplexNumber(this.real + that, this.imaginary);
+  def -(that:Double):ComplexNumber = new ComplexNumber(this.real - that, this.imaginary); // relate to additiveInverse?
+  def *(that:Double):ComplexNumber = new ComplexNumber(this.real * that, this.imaginary * that);
+  def /(that:Double):ComplexNumber = new ComplexNumber(this.real / that, this.imaginary / that); // relate to multiplicativeInverse?
+
+  // i'm sure there's a better way to do this...
+  override def equals(that:Any):Boolean = {
+    that match {                                 
+      case _:ComplexNumber => this.real == that.asInstanceOf[ComplexNumber].real && this.imaginary == that.asInstanceOf[ComplexNumber].imaginary; // WHY DOESNT THIS COMPILE GRR
```
Comment (tommysullivan) [inline lib/public/scala/complex-number.scala]: case ComplexNumber(real, imaginary) => this.real ==real && this.imaginary == imaginary

File: lib/public/scala/complex-number.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      case _:Byte => this.real == that && this.imaginary == 0;
```
Comment (tommysullivan) [inline lib/public/scala/complex-number.scala]: you shouldn't need all these underscores and colons you can just specify the type

File: lib/public/scala/complex-number.scala (hunk 1 above)
Comment (bob-bins) [inline lib/public/scala/complex-number.scala]: oops!

File: lib/public/scala/complex-number.scala (hunk 2 above)
Comment (bob-bins) [inline lib/public/scala/complex-number.scala]: I accidentally left in that comment (my screen was too small to show it without scrolling). The code worked after I added `asInstanceOf`. Your code is more concise, but it's giving me an error: `error: not found: value ComplexNumber
      case ComplexNumber(real, imaginary) => this.real == real && this.imaginary == imaginary;`

And if I leave in the `_:` I get the error: `error: '=>' expected but '(' found.
      case _:ComplexNumber(real, imaginary) => this.real == real && this.imaginary == imaginary;`

File: lib/public/scala/complex-number.scala (hunk 3 above)
Comment (bob-bins) [inline lib/public/scala/complex-number.scala]: That was the first thing I tried, but it gave me errors (similar to the first error in my first comment for the section above this one)

**Comment by bob-bins** [comment]
yep that worked

**Comment by bob-bins** [comment]
for the future:
1. combine RealVector and ComplexVector into a single class
2. override * operator for primitive numeric types, so `3 * ComplexNumber(1,1)` does not result in an error
3. find out how to make classes, objects, etc. in one file accessible to another in scala
4. create a proper testing method
5. find a way to convert scala to javascript so we can make use of this stuff for the website

**Comment by bob-bins** [comment]
The class was named `CoordVector` (for coordinate vector) because `Vector` is already a class in the standard library. I wasn't sure if it would cause issues to name ours the same.

I changed the input from `List` to `Vector` since `Vector` is more efficient on average, is also immutable, and I don't know of any advantages `List` would have in this context.

The way I combined the classes is I made any `Vector[Numeric]` input convert to a `Vector[ComplexNumber]`.

**Pull request merged**